### ISSUE TITLE
Struct fields and variant tags both go through deserialize_struct_field

### ISSUE THREAD

**Issue by dtolnay** [serde-rs/serde#722]
In `#[derive(Deserialize)]` we are generating effectively the same Deserialize implementation for deciding which struct field is next vs which variant we are looking at. As a result, both implementations go through Deserializer::deserialize_struct_field which is unexpected for Deserializer authors trying to implement `EnumVisitor::visit_variant`.

**Comment by dtolnay** [comment]
The workaround for binary formats is to go through U32Deserializer. Something like:

```rust
fn visit_variant_seed<V>(self, seed: V) -> Result<(V::Value, Self::Variant)>
    where V: de::DeserializeSeed
{
    // instead of `seed.deserialize(&mut *self.de)`
    let idx: u32 = Deserialize::deserialize(&mut *self.de)?;
    let val = seed.deserialize(idx.into_deserializer())?;
    Ok((val, /* ... */))
}
```

**Comment by dtolnay** [comment]
Fixed in #860.
